worlds/stardew_valley/logic.py

```diff
@@ -0,0 +1,1143 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Dict, Union, Optional, Iterable, Sized, Tuple, List
+
+from BaseClasses import CollectionState, ItemClassification
+from . import options
+from .bundle_data import BundleItem
+from .fish_data import all_fish_items
+from .game_item import FishItem
+from .items import all_items, Group, item_table
+from .options import StardewOptions
+
+MISSING_ITEM = "THIS ITEM IS MISSING"
+
+tool_materials = {
+    "Copper": 1,
+    "Iron": 2,
+    "Gold": 3,
+    "Iridium": 4
+}
+
+tool_prices = {
+    "Copper": 2000,
+    "Iron": 5000,
+    "Gold": 10000,
+    "Iridium": 25000
+}
+
+skill_level_per_season = {
+    "Spring": {
+        "Farming": 2,
+        "Fishing": 2,
+        "Foraging": 2,
+        "Mining": 2,
+        "Combat": 2,
+    },
+    "Summer": {
+        "Farming": 4,
+        "Fishing": 4,
+        "Foraging": 4,
+        "Mining": 4,
+        "Combat": 3,
+    },
+    "Fall": {
+        "Farming": 7,
+        "Fishing": 5,
+        "Foraging": 5,
+        "Mining": 5,
+        "Combat": 4,
+    },
+    "Winter": {
+        "Farming": 7,
+        "Fishing": 7,
+        "Foraging": 6,
+        "Mining": 7,
+        "Combat": 5,
+    },
+    "Year Two": {
+        "Farming": 10,
+        "Fishing": 10,
+        "Foraging": 10,
+        "Mining": 10,
+        "Combat": 10,
+    },
+}
+season_per_skill_level: Dict[Tuple[str, int], str] = {}
+season_per_total_level: Dict[int, str] = {}
+
+
+def initialize_season_per_skill_level():
+    current_level = {
+        "Farming": 0,
+        "Fishing": 0,
+        "Foraging": 0,
+        "Mining": 0,
+        "Combat": 0,
+    }
+    for season, skills in skill_level_per_season.items():
+        for skill, expected_level in skills.items():
+            for level_up in range(current_level[skill] + 1, expected_level + 1):
+                skill_level = (skill, level_up)
+                if skill_level not in season_per_skill_level:
+                    season_per_skill_level[skill_level] = season
+        level_up = 0
+        for level_up in range(level_up + 1, sum(skills.values()) + 1):
+            if level_up not in season_per_total_level:
+                season_per_total_level[level_up] = season
+
+
+initialize_season_per_skill_level()
+week_days = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
+
+
+class StardewRule:
```

Comment:
It's necessary because
- In the first beta, all the rule was made of lambda, like in other games. However, one cannot really know what's in a lambda without actually running it. It's really now convenient for debugging. 
- Some locations have such complicated logic that it would be extremely hard to reuse and maintain properly. For instances, with the bundle randomization, the rules changes based on the content of the bundle. I'd argue that having a rule for every possible item being already defined really simplified the implementation.  
- In retrospect, the logic framework we built is most likely once of the reason we could release so many features so fast, since it made it really easy to extend the logic.

Reply:
It's definitely an interesting approach. I'm more wondering if it's worth generalizing it, the .difficulty is hard to generalize though.

Reply:
It would definitely make logic debugging easier. I'm however not entirely satisfied of the simplification and the difficulty approximation, so that would probably need some improvement before generalizing.